### PR TITLE
fix: Add struct-wide RWMutex to metrics

### PR DESCRIPTION
**Pull request by simster7**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/metrics/metrics.go [hunk 1]
```diff
@@ -111,19 +114,23 @@ func (m *Metrics) WorkflowAdded(key string, phase v1alpha1.NodePhase) {
...
-	if fromPhase == toPhase || !hasKey {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if fromPhase == toPhase || !m.workflows[key] {
 		return
 	}
-	m.WorkflowDeleted(key, fromPhase)
-	m.WorkflowAdded(key, toPhase)
+	if _, ok := m.workflowsByPhase[fromPhase]; ok {
```
Comment (alexec) [inline workflow/metrics/metrics.go:123]: is this copy-and-paste? maybe make private func which does not lock?

**Comment by alexec** [comment]
it's like in Java when you add `synchronised` to everything

File: workflow/metrics/metrics.go (hunk 1 above)
Comment (simster7) [reply workflow/metrics/metrics.go:123]: Not quite copy-and-paste, there is some logic that would be redundant if we go with the private func approach; mainly, we will remove and readd the key from the map and perform an additional check unnecessarily. I actually think this is simpler and easier to understand.

However, I don't feel strongly and would gladly go with the private func approach if you think it's necessary.

Comment (alexec) [reply workflow/metrics/metrics.go:123]: we should just make sure that if we need to make changes in two places - we are aware of it

File: workflow/metrics/metrics.go [hunk 2]
```diff
@@ -77,6 +77,9 @@ func New(metricsConfig, telemetryConfig ServerConfig) *Metrics {
 }
 
 func (m *Metrics) allMetrics() []prometheus.Metric {
+	m.mutex.RLock()
```
Comment (sarabala1979) [inline workflow/metrics/metrics.go:80]: will it handle concurrent read and write on map?

Comment (sarabala1979) [reply workflow/metrics/metrics.go:80]: there is a synchronized map in golang.
https://golang.org/pkg/sync/#Map

**Pull request merged**